Add AI behavior and character.

LGM-Doyle opened the pull request:
Please use the following as the merge message 

**************\* Merge Message Cut Here ***********************************
Create AI Character composed of Behaviors to encompass variation 
between AIs.

Character solves the problem of having AI personalization added
piecemeal throughout the AI code.  Now each type of personalization is
encapsulated in a single Behavior.  The Behaviors are combined to form a
single unique Character per AI.

Refactor current variation in AI into two Behaviors, Aggression and 
EmpireIDBehavior to encompass AI variation due to the aggression setting 
and the empire id.

Add debugging bypass to randomized Character and Behavior.

The bypass can force individual AIs to have a specific
behavior and/or force all other AIs to have the same default behavior. 
********************\* End of Merge Message ****************************

I think this PR addresses issue #665.

It is a first step to more testable personalization of the AI.  If each behavior is kept orthogonal to the other behaviors then they can be tested in isolation, for functionality and performance.  This will keep testing manageable.

LGM-Doyle committed:
Expose GetOptionsDB().Get in AIInterface.

Add functions GetOptionsDBOptionsStr(), GetOptionsDBOptionsInt(),
GetOptionsDBOptionsBool() and GetOptionsDBOptionsDouble() in order to
allow forcing setting of AI behavior from options file.

LGM-Doyle committed:
Remove GetAIConfigStr() from AIInterface.

GetAIConfigStr() duplicates functionality from GetOptionsDBOptionStr(),
but only for the option 'ai-config'.

LGM-Doyle committed:
Add python test framework using pytest with tests for Character.py

The framework is simply a README file with examples of how to run
pytest.

The test for Character.py test that all of the combiners for multiple
behaviors work as expected. The combiners are min, max, preference and
all True.

It tests that indexing by behavior selects the correct behavior.
It tests that building a Character out of non-behaviors raises a
TypeError.

The tests for CharacterStrings check that make_aggression table make a
table of the correct length that is indexed by the aggression behavior
of a Character that contains Aggression and show UNKNOWN_VALUE for
Characters that don't contain aggression.

LGM-Doyle committed:
Add debugging bypass to randomized Character and Behavior.

Add a debugging bypass to force individual AIs to have a specific
behavior or to force all AIs to have the same behavior. If some AIs have
an individual behavior forced and some do not, then the remainder use
the .all setting.

Add three types of options to OptionsDB to force the value of aggression
and empire id behaviors.  The .force option is a bool indicating whether
the bypass is active.  The .all option is a default if a specific AI
does not have its own value.  The .AI_<number> option forces that
behavior to that value for AI_<number>.

The options are:
AI.config.behavior.<behavior name>.force  -- If true use the following
options to bypass the behavior
AI.config.behavior.<behavior name>.all    -- If present use this value
for all of the AIs not individually set
AI.config.behavior.<behavior name>.AI_1   -- Use for AI_1
   .
   .
   .
AI.config.behavior.<behavior name>.AI_40  -- Use for AI_40

, where <behavior name> is either aggression or empire-id.

An example of the code from persistent_config.xml to setup some AIs
follows.

<AI>
    <config>
      <behavior>
        <aggression>
          <force>1</force>
          <all>4</all>
          <AI_0>5</AI_0>
          <AI_1>4</AI_1>
          <AI_2>3</AI_2>
          <AI_3>2</AI_3>
          <AI_4>1</AI_4>
          <AI_5>0</AI_5>
        </aggression>
        <empire-id>
          <force>1</force>
          <AI_0>5</AI_0>
          <AI_1>4</AI_1>
          <AI_2>3</AI_2>
          <AI_3>2</AI_3>
          <AI_4>1</AI_4>
          <AI_5>0</AI_5>
        </empire-id>
      </behavior>
    </config>
</AI>

LGM-Doyle committed:
Adding Character, Behavior and CharacterString to AI.

Create AI Character composed of Behaviors to encompass variation between
AIs.

Character solves the problem of having AI personalization added
piecemeal throughout the AI code.  Now each type of personalization is
encapsulated in a single Behavior.  The Behaviors are combined to form a
single unique Character per AI.

CharacterStrings are tables of strings, typically UserStrings indexed by
Character.  They return single strings or lists of strings to be used to
communicate in a Character appropriate manner with the human player.

LGM-Doyle committed:
Create EmpireIDBehavior composed of all fo.empireID() % 2 references

Replace all AI personalization based on fo.empireID() % 2 (or 3) with a
single Behavior called EmpireIDBehavior.

Centralizes all related decisions and makes it easier to be
consistent in the future.

Makes it possible to turn EmpireIDBehaviors on/off from one location
independent of the actual empire id.

LGM-Doyle committed:
Create Aggression Behavior composed of all AIstate.aggression references

Replace all AI personalization based on aggression from many files into
a single Behavior called Aggression in one location.

Aggression centralizes all related decisions and makes it easier to be
consistent in the future.

Aggression will permit modifying for debug purposes all Aggression from
a single location.  For example, force all AIs to behave with the same
aggression.

LGM-Doyle committed:
Fixes for PR#656 Review

1. Fix inverted monster threat for exploration.
2. Fix extra underscore in CharacterStrings_test.py
3. Added comment on use of
   freeorion_debug.ide_tools.result.freeOrionAIInterface depending on
   removal of imports in __init__.py in its path.
4. Removed references to AIstate.aggression from DiplomaticCorp
5. Fix typo in Character refactor dt_tech -> dt_ech in AIResearch.py

LGM-Doyle committed:
Add stringtable options to bypass Aggression and EmpireID behaviors.

LGM-Doyle committed:
Removed non-functional key property from Character in Character.py

LGM-Doyle committed:
Move DiplomaticCorp.evaluate_diplomatic_attitude into Behavior.Agression

LGM-Doyle committed:
Move default/AI/Behavior to default/python/AI/Behavior.

LGM-Doyle committed:
Use Counter.most_common to count Behavior preferences.

Suggested by Morlic-fo in PR Review 677.

LGM-Doyle committed:
Make history analyzer candidate finder check character for aggression.

LGM-Doyle committed:
Move function get_behavior_bypass_value from AIstate to Character.py.

From PR review #677.

LGM-Doyle committed:
Refactor common code into EmpireIDBehavior._modulo_two_choice().

Due to review #677.

LGM-Doyle committed:
Comment changes in Character.py in response to review #677.

LGM-Doyle reviewed (commented):
I've made all of the changes as noted, except changing Behavior to Trait.  As mentioned I'm fine with that change, but I wanted to check before I did the work and found out I misunderstood your intent.

LGM-Doyle committed:
Add Behavior warship_adjusted_production_cost_exponent.

It pulls in an aggression mechanic from AIShipDesign.py

LGM-Doyle committed:
Rename directory default/python/AI/Behavior -> default/python/AI/Trait

First step of renaming AI Behavior to Trait.

As suggested in review #677.

LGM-Doyle committed:
Move Trait test code to tests directory.

LGM-Doyle committed:
Change Character.Behavior bypass into Character.Trait bypass.

As suggested in review #677.

LGM-Doyle committed:
Rename AI Character.Behavior to Character.Trait.

As suggested in review #677.

Morlic-fo reviewed (changes requested):
Seems fine now for the most part.

LGM-Doyle committed:
Rename AI/Trait module to AI/Character for review #677.

LGM-Doyle committed:
Miscellaneous changes for review #677 of AI Character code.
1. spelling sparce -> sparse
2. Improve comments
3. Fix inverted logic in ProductionAI.py
4. Replace function removed by rebase.

LGM-Doyle committed:
Use more generic descriptive names for CharacterStrings lists of strings.

LGM-Doyle committed:
Change AI CharacterString to a more natural function idiom.

Issue raised in review #677.

LGM-Doyle committed:
Use Character.warship_adjusted_production_cost_exponent() in ShipDesignAI.py

LGM-Doyle committed:
Add geometric_combiner to AI/Character for warship exponent.

From review #677.

LGM-Doyle committed:
Fix Character average_combiner to return 0 when passed empty input.

From review #677.

LGM-Doyle committed:
Rename Character package all lowercase for review #677.

LGM-Doyle committed:
Rename Character modules all lowercase for review #677.

LGM-Doyle reviewed (commented):
@Cjkjvfnby  and @Morlic-fo  thanks for the prompt and thorough review.  I've made changes or explained why I chose this implementation for all of your comments.

I've renamed the modules and packages as requested.

LGM-Doyle committed:
Fixes of averaging combiners for review #677

1. Fix PEP8 x is not None.
2. Cast length to float.
3. Guard against geometric mean of negative numbers.

LGM-Doyle committed:
PEP8 rename of character_strings_test.py

LGM-Doyle committed:
Rename trait_name_aggression to get_trait_name_aggression in AI code.

From review of #677.

LGM-Doyle committed:
Add tests to character_strings_test for get_trait_name_aggression.

LGM-Doyle committed:
Rename functions in character_strings_modules to make them private.

In response to review #677.

LGM-Doyle committed:
Expand file comments in character_strings_module.py.

The expanded comments clarify the types of public functions.

LGM-Doyle committed:
Fix Character.get_trait(x) to return None if it has not trait x.

This is the documented behavior.

Further fixes were required in character_strings_module to handle the
possible None results.
